extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java

```diff
@@ -402,16 +407,217 @@ private void loadSubtitles() {
                         final String ext = url.substring(url.lastIndexOf(".") + 1);
                         final MediaFormat fmt = MediaFormat.getFromSuffix(ext);
-                        if (fmt != null && !isNullOrEmpty(languageCode))
-                            subtitles.add(new SubtitlesStream(fmt, languageCode, url, false));
+                        if (fmt != null && !isNullOrEmpty(languageCode)) {
+                            subtitles.add(new SubtitlesStream(url, fmt, languageCode, false));
+                        }
                     }
                 }
-            } catch (Exception e) {
-                // ignore all exceptions
+            } catch (final Exception ignored) {
+                // Ignore all exceptions
             }
         }
     }
 
+    private void extractLiveVideoStreams() throws ParsingException {
+        try {
+            final JsonArray streamingPlaylists = json.getArray(STREAMING_PLAYLISTS);
+            for (final Object s : streamingPlaylists) {
+                if (!(s instanceof JsonObject)) {
+                    continue;
+                }
+                final JsonObject stream = (JsonObject) s;
+                // Don't use the containsSimilarStream method because it will always return false
+                // so if there are multiples HLS URLs returned, only the first will be extracted in
+                // this case.
+                videoStreams.add(new VideoStream(String.valueOf(stream.getInt("id", 0)),
+                        stream.getString(PLAYLIST_URL, EMPTY_STRING), true, MediaFormat.MPEG_4,
+                        DeliveryMethod.HLS, EMPTY_STRING, false, null));
+            }
+        } catch (final Exception e) {
+            throw new ParsingException("Could not get video streams", e);
+        }
+    }
+
+    private void getStreams() throws ParsingException {
+        // Progressive streams
+        getStreamsFromArray(json.getArray(FILES), EMPTY_STRING);
+
+        // HLS streams
+        try {
+            final JsonArray streamingPlaylists = json.getArray(STREAMING_PLAYLISTS);
+            for (final Object p : streamingPlaylists) {
+                if (!(p instanceof JsonObject)) {
+                    continue;
+                }
+                final JsonObject playlist = (JsonObject) p;
+                final String playlistUrl = playlist.getString(PLAYLIST_URL);
+                getStreamsFromArray(playlist.getArray(FILES), playlistUrl);
+            }
+        } catch (final Exception e) {
+            throw new ParsingException("Could not get streams", e);
+        }
+    }
+
+    private void getStreamsFromArray(@Nonnull final JsonArray streams,
+                                     final String playlistUrl) throws ParsingException {
+        try {
+            /*
+            Starting with version 3.4.0 of PeerTube, HLS playlist of stream resolutions contain the
```

Comment:
Do we really need to implement this for Peertube <3.4.0 (which is outdated) in the first place?
Currently only 20-30% of the Peertube instances run such an version.

Reply:
Yes, we need, otherwise we would get wrong streaming URLs which will not work. Why not keeping support of outdated instances, at least for some months?

> Currently only 20-30% of the Peertube instances run such an version.

Also, where did you got this?

Reply:
20% is quite a lot

Reply:
> Also, where did you got this?

https://instances.joinpeertube.org/instances